### PR TITLE
feat: pausable protocol token

### PR DESCRIPTION
**Pull request by 0xJabberwock**
Closes HAI-318

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -67,7 +67,7 @@ jobs:
 
       - name: Run tests
         shell: bash
-        run: yarn test --nmt _Bytecode
+        run: yarn test
```
Comment (wei3erHase) [inline .github/workflows/test.yml:70]: 👀

File: script/Common.s.sol
Comment (wei3erHase) [inline script/Common.s.sol:1]: why so many changes in this file?

File: script/Common.s.sol [hunk 2]
```diff
@@ -11,142 +11,15 @@ abstract contract Common is Contracts, Params {
...
-    return governor != deployer && governor != address(0);
-  }
-
   function deployTokens() public updateParams {
     systemCoin = new SystemCoin('HAI Index Token', 'HAI');
     protocolToken = new ProtocolToken('Protocol Token', 'KITE');
   }
 
+  function _setupTokens() internal {
+    protocolToken.pause();
```
Comment (wei3erHase) [inline script/Common.s.sol]: `protocolToken` should be paused on constructor, not in a following action

File: script/Deploy.s.sol [hunk 3]
```diff
@@ -20,6 +20,7 @@ abstract contract Deploy is Common, Script {
 
     // Deploy tokens used to setup the environment
     deployTokens();
+    _setupTokens();
```
Comment (wei3erHase) [inline script/Deploy.s.sol]: 👀 ?

File: src/contracts/tokens/ProtocolToken.sol [hunk 4]
```diff
@@ -1,20 +1,27 @@
...
 /**
  * @title  ProtocolToken
  * @notice This contract represents the protocol ERC20Votes token to be used for governance purposes
  */
-contract ProtocolToken is ERC20, ERC20Permit, ERC20Votes, Authorizable, IProtocolToken {
+contract ProtocolToken is ERC20, ERC20Permit, ERC20Votes, ERC20Pausable, Authorizable, IProtocolToken {
+  // --- Data ---
+
+  /// @inheritdoc IProtocolToken
+  bool public notPausable;
```
Comment (wei3erHase) [inline src/contracts/tokens/ProtocolToken.sol]: what? doesn't Pausable already inherits a `paused` var?

File: src/contracts/tokens/ProtocolToken.sol [hunk 5]
```diff
@@ -38,9 +45,21 @@ contract ProtocolToken is ERC20, ERC20Permit, ERC20Votes, Authorizable, IProtoco
     _burn(msg.sender, _wad);
   }
 
+  /// @inheritdoc IProtocolToken
+  function pause() external isAuthorized {
+    if (notPausable) revert ProtocolToken_NotPausable();
+    notPausable = true;
+    _pause();
+  }
```
Comment (wei3erHase) [inline src/contracts/tokens/ProtocolToken.sol]: `- not pausable again` means not adding this method at all

File: src/contracts/tokens/ProtocolToken.sol [hunk 6]
```diff
@@ -38,9 +45,21 @@ contract ProtocolToken is ERC20, ERC20Permit, ERC20Votes, Authorizable, IProtoco
...
+
+  /// @inheritdoc IProtocolToken
+  function unpause() external isAuthorized {
+    _unpause();
+  }
```
Comment (wei3erHase) [inline src/contracts/tokens/ProtocolToken.sol:48]: add a 'when paused' clause 🙏 
(only if `_unpause()` doesn't have one)

File: src/contracts/tokens/ProtocolToken.sol [hunk 7]
```diff
@@ -38,9 +45,21 @@ contract ProtocolToken is ERC20, ERC20Permit, ERC20Votes, Authorizable, IProtoco
...
+
   // --- Overrides ---
 
-  function _update(address _from, address _to, uint256 _value) internal override(ERC20, ERC20Votes) {
+  function _update(address _from, address _to, uint256 _value) internal override(ERC20, ERC20Votes, ERC20Pausable) {
```
Comment (wei3erHase) [inline src/contracts/tokens/ProtocolToken.sol:52]: q: what's `_update` in Pausable contract?

File: src/contracts/tokens/TokenDistributor.sol [hunk 8]
```diff
@@ -73,11 +71,13 @@ contract TokenDistributor is Authorizable, ITokenDistributor {
   /// @inheritdoc ITokenDistributor
   function sweep(address _sweepReceiver) external override isAuthorized {
     if (block.timestamp <= claimPeriodEnd) revert TokenDistributor_ClaimPeriodNotEnded();
-    uint256 _balance = token.balanceOf(address(this)).assertGt(0);
 
-    token.safeTransfer(_sweepReceiver, _balance);
+    uint256 _totalClaimable = totalClaimable.assertNonNull();
+    delete totalClaimable;
+
+    token.mint(_sweepReceiver, _totalClaimable);
```
Comment (wei3erHase) [inline src/contracts/tokens/TokenDistributor.sol:78]: i'm ok with it 👍

File: src/interfaces/tokens/IProtocolToken.sol [hunk 9]
```diff
@@ -4,9 +4,25 @@ pragma solidity 0.8.20;
...
+  /// @notice Throws when trying to pause the token a second time
+  error ProtocolToken_NotPausable();
+
+  // --- Data ---
+
+  /**
+   * @notice Pausability status of the token
+   * @return _notPausable Whether the token is pausable or not
+   */
+  function notPausable() external view returns (bool _notPausable);
```
Comment (wei3erHase) [inline src/interfaces/tokens/IProtocolToken.sol]: whhattttt? why using `pausable`?

File: src/interfaces/tokens/IProtocolToken.sol [hunk 10]
```diff
@@ -20,4 +36,16 @@ interface IProtocolToken is IERC20Metadata, IERC20Permit, IVotes, IAuthorizable
    * @param _amount Amount of tokens to burn [wad]
    */
   function burn(uint256 _amount) external;
+
+  /**
+   * @notice Pause the token transfers, minting and burning
+   * @dev    Only authorized addresses can pause the token
+   */
+  function pause() external;
```
Comment (wei3erHase) [inline src/interfaces/tokens/IProtocolToken.sol]: rm 🙏

File: test/e2e/Common.t.sol [hunk 11]
```diff
@@ -58,6 +58,13 @@ contract DeployForTest is TestParams, Deploy {
     collateralTypes.push('TKN-C');
     collateralTypes.push('TKN-8D');
   }
+
+  function setupPostEnvironment() public virtual override updateParams {
+    for (uint256 i = 0; i < collateralTypes.length; i++) {
+      bytes32 _cType = collateralTypes[i];
+      taxCollector.taxSingle(_cType);
+    }
+  }
```
Comment (wei3erHase) [inline test/e2e/Common.t.sol:67]: unnecessary changes?

File: test/unit/HaiSafeManager.t.sol [hunk 12]
```diff
@@ -14,7 +14,10 @@ abstract contract Base is HaiTest {
 
   function setUp() public virtual {
     vm.startPrank(deployer);
+
     safeManager = new HaiSafeManagerForTest(address(safeEngine));
+    label(address(safeManager), 'HaiSafeManager');
```
Comment (wei3erHase) [inline test/unit/HaiSafeManager.t.sol:19]: why removing label on previous tests but adding one here?

File: test/unit/tokens/ProtocolToken.t.sol
Comment (wei3erHase) [inline test/unit/tokens/ProtocolToken.t.sol:1]: good 👍

**Review by wei3erHase — CHANGES_REQUESTED** [review]
good approach! i'd prefer making as less changes as possible across repo, but they seem quite trivial so i'm ok, yet the script changes seem unnecessary, wdyt?
as for the Paused, blz remove `pause()` method, and make the contract paused at constructor, also, there's no need for `notPausable` variable

File: .github/workflows/test.yml (hunk 1 above)
Comment (wei3erHase) [reply .github/workflows/test.yml:70]: ahh you disabled both deployment tests completely, hmmmmmm 🤔

File: src/contracts/tokens/ProtocolToken.sol (hunk 6 above)
Comment (0xJabberwock) [reply src/contracts/tokens/ProtocolToken.sol:48]: Indeed, `_unpause()` does include a `whenPaused` modifier.

File: src/contracts/tokens/ProtocolToken.sol (hunk 7 above)
Comment (0xJabberwock) [reply src/contracts/tokens/ProtocolToken.sol:52]: `_update()` is overridden in `ERC20Pausable` so it includes the `whenNotPaused` modifier, consequently affecting the behaviour of `_mint()`, `_transfer()` and `_burn()` methods.

File: test/unit/HaiSafeManager.t.sol (hunk 12 above)
Comment (0xJabberwock) [reply test/unit/HaiSafeManager.t.sol:19]: In unit tests, the labeling after main contract deployment was removed only in those cases where it was redundant (i.e., when contract name matches the desired label).

As seen in Foundry traces, contracts deployed using the `new` keyword adopt their name as label in an automatic manner.

In the highlighted case, the `safeManager` address is initially labeled as `HaiSafeManagerForTest`, which doesn't coincide with the labeling approach we've been following. Hence the line addition.

File: test/e2e/Common.t.sol (hunk 11 above)
Comment (0xJabberwock) [reply test/e2e/Common.t.sol:67]: I think it makes the for-test deployment routine tidier.

File: script/Common.s.sol
Comment (0xJabberwock) [reply script/Common.s.sol:1]: All of them are for a matter of reordering so it coincides with the actual deployment sequence, except for the addition of `_setupTokens()` and the change in `deployTokenDistributor()`.

File: .github/workflows/test.yml (hunk 1 above)
Comment (0xJabberwock) [reply .github/workflows/test.yml:70]: They started to fail after testing for the protocol token pausability in `CommonDeploymentTest`.

File: src/contracts/tokens/ProtocolToken.sol (hunk 4 above)
Comment (0xJabberwock) [reply src/contracts/tokens/ProtocolToken.sol]: A `_paused` boolean exists, but cannot be _directly_ overwritten outside `Pausable` because it is `private`.

File: src/interfaces/tokens/IProtocolToken.sol (hunk 9 above)
Comment (0xJabberwock) [reply src/interfaces/tokens/IProtocolToken.sol]: It was used to introduce a pausability status, in addition to the pause state. However, as you pointed out, it is not the best approach here as we know upfront when it should no longer be pausable.

File: script/Common.s.sol
Comment (wei3erHase) [reply script/Common.s.sol:1]: hmmm understandable, BUT, makes unnecesary changes on test suite, like, i must now review line by line if it's accurate, imo is not needed at this stage, and makes the PR smaller

File: src/contracts/tokens/ProtocolToken.sol (hunk 7 above)
Comment (wei3erHase) [reply src/contracts/tokens/ProtocolToken.sol:52]: ALSO MINT?!

File: test/single/SurplusAuctionHouse.t.sol [hunk 13]
```diff
@@ -152,7 +152,10 @@ contract SingleBurningSurplusAuctionHouseTest is DSTest {
 
     safeEngine.createUnbackedDebt(address(this), address(this), 1000 ether);
 
+    protocolToken.unpause();
```
Comment (wei3erHase) [inline test/single/SurplusAuctionHouse.t.sol:155]: why is this not in the setup?

File: .github/workflows/test.yml (hunk 1 above)
Comment (wei3erHase) [reply .github/workflows/test.yml:70]: hmm here's the catch, i'd modify `test:local` to ignore `_Bytecode`, but making them `abstract` means that if i wanna run them locally, i cannot w/o modifying the codebase.

yet, since we use that bash sh*t, we cannot do `yarn test:local --mc E2E`, so issa pain to have non-working tests, i'm on a crossover here 🤔

File: src/contracts/tokens/ProtocolToken.sol [hunk 14]
```diff
@@ -38,9 +42,14 @@ contract ProtocolToken is ERC20, ERC20Permit, ERC20Votes, Authorizable, IProtoco
...
     super._update(_from, _to, _value);
```
Comment (wei3erHase) [inline src/contracts/tokens/ProtocolToken.sol]:
```suggestion
    ERC20Votes._update(_from, _to, _value);
```

File: test/e2e/Deploy.t.sol [hunk 15]
```diff
@@ -124,9 +126,38 @@ abstract contract CommonDeploymentTest is HaiTest, Deploy {
...
+
+  function test_ProtocolToken_Pausable(uint256 _wad) public {
+    vm.assume(_wad <= type(uint208).max);
+
+    vm.startPrank(deployer);
+    protocolToken.approve(governor, _wad);
+    changePrank(governor);
+
+    vm.expectRevert(Pausable.EnforcedPause.selector);
+    protocolToken.mint(governor, _wad);
```
Comment (wei3erHase) [inline test/e2e/Deploy.t.sol:142]: ⚠️  token should be minted to vote to be unpaused

File: src/contracts/tokens/ProtocolToken.sol (hunk 4 above)
Comment (wei3erHase) [reply src/contracts/tokens/ProtocolToken.sol]: lol, there must be a reason for that 🤔 
that's their design enforcing the dev to use their methods

File: test/single/SurplusAuctionHouse.t.sol (hunk 13 above)
Comment (0xJabberwock) [reply test/single/SurplusAuctionHouse.t.sol:155]: It is in the `setUp()`. 😉

File: src/contracts/tokens/ProtocolToken.sol (hunk 14 above)
Comment (0xJabberwock) [reply src/contracts/tokens/ProtocolToken.sol]: Then there's no reason to inherit `ERC20Pausable`, but `Pausable`.

Comment (wei3erHase) [reply src/contracts/tokens/ProtocolToken.sol]: OHHHHH SHI, they changed the implementation to this sh*t! 😡 💢

Comment (wei3erHase) [reply src/contracts/tokens/ProtocolToken.sol]: wdyt about:
```suggestion
    // Override ERC20Pausable when minting new tokens
    if (_from == 0) ERC20._update(_from, _to, _value);
    else ERC20Pausable._update(_from, _to, _value);
```

i think that with current OZ setup, at SOME point we're gonna have to be making sth like this, since there's no `_beforeTokenTransfer` hook anymor

File: .github/workflows/test.yml (hunk 1 above)
Comment (0xJabberwock) [reply .github/workflows/test.yml:70]: Most tests whose fork environment consists of the contracts fetched from the latest deployment fail.

Specifically in Mainnet, all fail but 3.

Specifically in Goerli, none `_Bytecode` ones pass, but `_Params`, `_CParams` and `_Auth` ones do (except for the modified `test_ProtocolToken_Auth`). Of course, the newly added `test_ProtocolToken_Pausable` also fails.

Will the contracts be redeployed on Goerli? Soon? If not, selectively ignoring the test cases that start to fail within each PR doesn't seem like an elegant solution. 😕

By the way, I can run `yarn test:local --mc E2E` and it executes well.

File: test/single/SurplusAuctionHouse.t.sol (hunk 13 above)
Comment (wei3erHase) [reply test/single/SurplusAuctionHouse.t.sol:155]: oh, we have multiple setups, oki :)

**Review by wei3erHase — APPROVED** [review]
great job pal! looks good to me! 🚀
